File: R/sampling_frame.R
#' Sampling Frame for fMRI Temporal Structure
#'
#' Creates and manipulates sampling frame objects that represent the temporal
#' structure of fMRI datasets. A sampling frame encapsulates run lengths,
#' repetition time (TR), and provides derived temporal properties.
#'
#' @param run_length A numeric vector of run lengths (number of timepoints per run)
#' @param TR Repetition time in seconds
#' @return A sampling_frame object
#' @export
sampling_frame <- function(run_length, TR) {
  assertthat::assert_that(is.numeric(run_length), all(run_length > 0))
  assertthat::assert_that(is.numeric(TR), length(TR) == 1, TR > 0)
  
  structure(
    list(
      run_length = run_length,
      TR = TR
    ),
    class = "sampling_frame"
  )
}

#' Test if Object is a Sampling Frame
#'
#' This function tests whether an object is of class 'sampling_frame'.
#'
#' @param x An object to test
#' @return TRUE if x is a sampling_frame object, FALSE otherwise
#' @export
is.sampling_frame <- function(x) {
  inherits(x, "sampling_frame")
}

#' @export
get_TR.sampling_frame <- function(x, ...) {
  x$TR
}

#' @export
get_run_lengths.sampling_frame <- function(x, ...) {
  x$run_length
}

#' @export
n_runs.sampling_frame <- function(x, ...) {
  length(x$run_length)
}

#' @export
n_timepoints.sampling_frame <- function(x, ...) {
  sum(x$run_length)
}

#' @export
blocklens.sampling_frame <- function(x, ...) {
  x$run_length
}

#' @export
blockids.sampling_frame <- function(x, ...) {
  rep(1:length(x$run_length), times = x$run_length)
}

#' @export
samples.sampling_frame <- function(x, ...) {
  1:sum(x$run_length)
}

#' @export
global_onsets.sampling_frame <- function(x, ...) {
  if (length(x$run_length) == 1) {
    return(1)
  }
  c(1, cumsum(x$run_length[-length(x$run_length)]) + 1)
}

#' @export
get_total_duration.sampling_frame <- function(x, ...) {
  sum(x$run_length) * x$TR
}

#' @export
get_run_duration.sampling_frame <- function(x, ...) {
  x$run_length * x$TR
}

#' @export
print.sampling_frame <- function(x, ...) {
  cat("Sampling Frame:\n")
  cat("  TR:", x$TR, "seconds\n")
  cat("  Runs:", length(x$run_length), "\n")
  cat("  Run lengths:", paste(x$run_length, collapse = ", "), "\n")
  cat("  Total timepoints:", sum(x$run_length), "\n")
  cat("  Total duration:", get_total_duration(x), "seconds\n")
  invisible(x)
} 

File: python/src/fmridataset/sampling_frame.py
"""SamplingFrame: temporal structure representation for fMRI acquisitions.

Direct port of ``fmrihrf::sampling_frame`` plus the adapter methods from
``R/sampling_frame_adapters.R``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence

import numpy as np
from numpy.typing import NDArray


@dataclass(frozen=True)
class SamplingFrame:
    """Immutable description of the temporal structure of an fMRI acquisition.

    Parameters
    ----------
    blocklens : tuple[int, ...]
        Number of time-points (volumes) in each run/block.
    TR : float
        Repetition time in seconds.
    """

    blocklens: tuple[int, ...]
    TR: float  # noqa: N815 – matches R naming

    def __post_init__(self) -> None:
        if not self.blocklens:
            raise ValueError("blocklens must be non-empty")
        if any(b <= 0 for b in self.blocklens):
            raise ValueError("all blocklens must be positive integers")
        if self.TR <= 0:
            raise ValueError("TR must be positive")

    # ------------------------------------------------------------------
    # Convenience constructor
    # ------------------------------------------------------------------

    @classmethod
    def create(
        cls,
        blocklens: Sequence[int],
        TR: float,  # noqa: N803
    ) -> SamplingFrame:
        """Create a SamplingFrame from a sequence of block lengths and TR."""
        return cls(blocklens=tuple(int(b) for b in blocklens), TR=float(TR))

    # ------------------------------------------------------------------
    # Properties – mirror R generic methods
    # ------------------------------------------------------------------

    @property
    def n_runs(self) -> int:
        """Number of runs/blocks."""
        return len(self.blocklens)

    @property
    def n_timepoints(self) -> int:
        """Total number of time-points across all runs."""
        return sum(self.blocklens)

    @property
    def blockids(self) -> NDArray[np.intp]:
        """1-based run id for every time-point (matches R ``rep(seq_along(...))``).

        Returns an array of length ``n_timepoints`` where each element is the
        1-based index of the run that time-point belongs to.
        """
        return np.repeat(
            np.arange(1, self.n_runs + 1),
            self.blocklens,
        )

    @property
    def samples(self) -> NDArray[np.intp]:
        """1-based sample indices (1 .. n_timepoints)."""
        return np.arange(1, self.n_timepoints + 1)

    @property
    def run_durations(self) -> NDArray[np.float64]:
        """Duration of each run in seconds (blocklens * TR)."""
        return np.array(self.blocklens, dtype=np.float64) * self.TR

    @property
    def total_duration(self) -> float:
        """Total acquisition duration in seconds."""
        return float(np.sum(self.run_durations))

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def run_indices(self, run: int) -> NDArray[np.intp]:
        """Return 0-based row indices for the given 1-based *run* number."""
        if run < 1 or run > self.n_runs:
            raise ValueError(
                f"run must be between 1 and {self.n_runs}, got {run}"
            )
        start = sum(self.blocklens[: run - 1])
        end = start + self.blocklens[run - 1]
        return np.arange(start, end)

    def __repr__(self) -> str:
        return (
            f"SamplingFrame(n_runs={self.n_runs}, "
            f"n_timepoints={self.n_timepoints}, TR={self.TR})"
        )
